pkg/zstdchunked/storage.go

```diff
@@ -0,0 +1,977 @@
+	if _, _, err := unix.Syscall6(unix.SYS_UTIMENSAT, uintptr(fd), 0, uintptr(unsafe.Pointer(&ts[0])), 0, 0, 0); err != 0 && err != unix.ENOSYS {
+		return err
+	}
+
+	// Files must be already created with the correct mode, but we must make sure the setuid/setgid
+	// is set.
+	if t == tar.TypeDir || mode&(os.ModeSetuid|os.ModeSetgid) != 0 {
+		if err := unix.Fchmod(fd, uint32(mode)); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func openFileUnderRoot(name, root string, dirfd int, flags uint64, mode os.FileMode) (*os.File, error) {
+	how := unix.OpenHow{
+		Flags:   flags,
+		Mode:    uint64(mode & 07777),
+		Resolve: unix.RESOLVE_IN_ROOT,
+	}
+
+	fd, err := unix.Openat2(dirfd, name, &how)
+	if err == nil {
+		return os.NewFile(uintptr(fd), name), nil
+	}
+
+	path, err := securejoin.SecureJoin(root, name)
+	if err != nil {
+		return nil, err
+	}
+
+	return os.OpenFile(path, int(flags), mode)
+}
+
+func createFileFromZstdStream(dest string, dirfd int, reader io.Reader, missingDirsMode, mode os.FileMode, metadata *FileMetadata, options *archive.TarOptions) (err error) {
+	file, err := openFileUnderRoot(metadata.Name, dest, dirfd, newFileFlags, mode)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err2 := file.Close()
+		if err == nil {
+			err = err2
+		}
+	}()
+
+	z, err := zstd.NewReader(reader)
+	if err != nil {
+		return err
+	}
+	defer z.Close()
+
+	digester := digest.Canonical.Digester()
+	checksum := digester.Hash()
+	_, err = z.WriteTo(io.MultiWriter(file, checksum))
+	if err != nil {
+		return err
+	}
+	manifestChecksum, err := digest.Parse(metadata.Digest)
+	if err != nil {
+		return err
+	}
+	if digester.Digest() != manifestChecksum {
+		return fmt.Errorf("checksum mismatch for %q", dest)
+	}
+	return setFileAttrs(file, mode, metadata, options)
+}
+
+func storeMissingFiles(stream io.Reader, contentType string, dest string, dirfd int, missingChunks []missingChunk, missingDirsMode os.FileMode, options *archive.TarOptions) error {
+	mediaType, params, err := mime.ParseMediaType(contentType)
+	if err != nil {
+		return err
+	}
+	if !strings.HasPrefix(mediaType, "multipart/") {
+		if len(missingChunks) != 1 {
+			return errors.Errorf("invalid response, no multipart for multiple chunks requested")
+		}
+
+		for _, mf := range missingChunks[0].Files {
+			if mf.Gap > 0 {
+				limitReader := io.LimitReader(stream, mf.Gap)
+				_, err := io.Copy(ioutil.Discard, limitReader)
+				if err != nil {
+					return err
+				}
+				continue
+			}
+			limitReader := io.LimitReader(stream, mf.Length())
+
+			if err := createFileFromZstdStream(dest, dirfd, limitReader, missingDirsMode, os.FileMode(mf.File.Mode), mf.File, options); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+
+	// multipart response
+	boundary, found := params["boundary"]
+	if !found {
+		return errors.Errorf("boundary param not found")
+	}
+
+	buffered := makeBufferedNetworkReader(stream, 64, 16384)
+	defer buffered.Close()
```

Comment:
What does this buffered network reader do? Isn’t it just an expensive way to maintain kernel-side socket buffers?

It is non-obvious to me why it is worth the complexity — and if it is, why aren’t we using it elsewhere / why isn’t it a part of the standard library? What’s special about this case?

Reply:
not sure if it can be useful elsewhere, but it makes a significant difference here so we can keep fetching data while it is being processed at the same time.  From what I could see this was a bottleneck when pulling an image from a remote registry.